File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.cpp
class Solution {
public:
    int sumSubarrayMins(vector<int>& arr) {
        int n = arr.size();
        stack<int> st1, st2;
        vector<int> next(n), prev(n);
        long long ans = 0, mod = 1e9 + 7;

        for(int i = 0; i<n; i++) {
            next[i] = n - i - 1;
            prev[i] = i;
        }

        for(int i = 0; i<n; i++) {
            while(!st1.empty() && arr[st1.top()] > arr[i]) {
                next[st1.top()] = i - st1.top() - 1;
                st1.pop();
            }
            st1.push(i);
        }

        
        for(int i = n - 1; i>=0; i--) {
            while(!st2.empty() && arr[st2.top()] >= arr[i]) {
                prev[st2.top()] = st2.top() - i - 1;
                st2.pop();
            }
            st2.push(i);
        }

        for(int i = 0; i<n; i++) {
            long long pro = ((next[i] + 1) * (prev[i] + 1)) % mod;
            pro = (arr[i]*pro) % mod;
            ans = (ans + pro) % mod;
        }

        return (int) ans;
    }
};

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.py
class Solution(object):
    def sumSubarrayMins(self, arr):
        n = len(arr)
        st1 = []
        st2 = []
        next = [0] * n
        prev = [0] * n
        ans = 0
        mod = 10**9 + 7

        for i in range(n):
            next[i] = n - i - 1
            prev[i] = i

        for i in range(n):
            while st1 and arr[st1[-1]] > arr[i]:
                next[st1[-1]] = i - st1[-1] - 1
                st1.pop()
            st1.append(i)

        for i in range(n - 1, -1, -1):
            while st2 and arr[st2[-1]] >= arr[i]:
                prev[st2[-1]] = st2[-1] - i - 1
                st2.pop()
            st2.append(i)

        for i in range(n):
            pro = ((next[i] + 1) * (prev[i] + 1)) % mod
            pro = (arr[i] * pro) % mod
            ans = (ans + pro) % mod

        return int(ans)
        